flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -224,54 +565,48 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
     }
 
     rewriter.eraseOp(doLoop);
+
+    if (shouldSkipInnerLoops)
+      processNotPerfectlyNestedLoops(rewriter, mapper, ompLoopNest);
+
     return mlir::success();
   }
 
 private:
-  /// Collect the list of values used inside the loop but defined outside of it.
-  /// The first item in the returned list is always the loop's induction
-  /// variable.
-  void collectLoopLiveIns(fir::DoLoopOp doLoop,
-                          llvm::SmallVectorImpl<mlir::Value> &liveIns) const {
-    // Given an operation `op`, this lambda returns true if `op`'s operand is
-    // ultimately the loop's induction variable. Detecting this helps finding
-    // the live-in value corresponding to the induction variable in case the
-    // induction variable is indirectly used in the loop (e.g. throught a cast
-    // op).
-    std::function<bool(mlir::Operation * op)> isIndVarUltimateOperand =
-        [&](mlir::Operation *op) {
-          if (auto storeOp = mlir::dyn_cast_if_present<fir::StoreOp>(op)) {
-            return (storeOp.getValue() == doLoop.getInductionVar()) ||
-                   isIndVarUltimateOperand(storeOp.getValue().getDefiningOp());
-          }
-
-          if (auto convertOp = mlir::dyn_cast_if_present<fir::ConvertOp>(op)) {
-            return convertOp.getOperand() == doLoop.getInductionVar() ||
-                   isIndVarUltimateOperand(
-                       convertOp.getValue().getDefiningOp());
-          }
-
-          return false;
-        };
-
-    llvm::SmallDenseSet<mlir::Value> seenValues;
-    llvm::SmallDenseSet<mlir::Operation *> seenOps;
-
-    mlir::visitUsedValuesDefinedAbove(
-        doLoop.getRegion(), [&](mlir::OpOperand *operand) {
-          if (!seenValues.insert(operand->get()).second)
-            return;
-
-          mlir::Operation *definingOp = operand->get().getDefiningOp();
-          // We want to collect ops corresponding to live-ins only once.
-          if (definingOp && !seenOps.insert(definingOp).second)
-            return;
-
-          liveIns.push_back(operand->get());
-
-          if (isIndVarUltimateOperand(operand->getOwner()))
-            std::swap(*liveIns.begin(), *liveIns.rbegin());
-        });
+  /// For loops that are not perfectly nested, we need to do 2 things:
+  /// 1. Privatize the memory definition/allocation of their iteration variables
+  ///    since these would cross the boundaries of the target/parallel region
+  ///    after mapping to OpenMP.
```

Comment:
Everything else will be either `shared` by default in case we are mapping to host or mapped using a `omp.map.info` (see the case to `genMapInfoOpForLiveIn(...)`. This is at least the behavior until we support locality specs for `do concurrent`.

For loop iteration variables however, don't we need private copies for each thread is a parallel team?

Did I get something wrong here?

Reply:
> For loop iteration variables however, don't we need private copies for each thread is a parallel team?

The user will probably want it privatized, but possibly also many other variables that have race conditions when executed in parallel. I think it is a slippery slope if the compiler starts to heuristically determine to implicitly privatize some variables, but not others. For instance:
```fortran
integer :: i, j

do concurrent (j = 1:42)
  i = 1
  do while (i < 11)
    print *, i
    i = i + 1
  end do
end do
```
is a race condition but
```fortran
integer :: i, j

do concurrent (j = 1:42)
  do i = 1, 10
    print *, i
  end do
end do
```
is not?

https://flang.llvm.org/docs/DoConcurrent.html aludes to *localization* rules:
```fortran
  DO CONCURRENT (J=1:N)
    TMP = A(J) + B(J)
    C(J) = TMP
  END DO
  ! And TMP is undefined afterwards
```
Here, is seems that `TMP` must be privatized as well. However, it seems to be unimplementable with
```fortran
  DO CONCURRENT (J=1:N)
    T(IX(J)) = A(J) + B(J)
    C(J) = T(IY(J))
  END DO
```
as documented there.

If it is the defined semantics of `-fdo-concurrent-parallel` for localization to happen only in nested loop variables, it needs to be really well documented. Otherwise, I think it is reasonable to require users to nest the loop inside a `BLOCK`.

Reply:
You are correct. Thanks for catching that. I undid that change. We can think about dependency analyses that we can implement to selectively privatize IV vars and other stuff as you suggested on Teams.